DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,768 @@
+When overriding parameters, there needs to be an exact match however.
+
+```D
+class A {
+    A foo(A param) {
+        return param;
+    }
+}
+
+class B {
+    // return type B is allowed since B < A
+    // parameter type Object is not allowed desipte A < Object
+    override B foo(Object param) {
+        assert(0);
+    }
+}
+```
+
+The addition of a bottom type doesn't change these rules: an override of a function returning a `T` may have return type `noreturn` (since `noreturn` is a subtype of every `T`) but a parameter of type `noreturn` may not be replaced by `T` (despite `T` being a supertype of `noreturn`).
+
+### Grammar changes
+Throw stratements become expressions so they can be used in lambda functions.
+
+```diff
+NonEmptyStatementNoCaseNoDefault:
+- ThrowStatement
+
+- ThrowStatement:
+-    throw Expression ;
+
+UnaryExpression:
++ ThrowExpression
+
++ ThrowExpression:
++    throw Expression
+```
+
+### Limitations
+While C functions can be declared using the noreturn type because of the implicit conversion and no name mangling, C++ function pointers have the return type in their mangling.
+When interfacing with C++ functions, changing the return type to `noreturn` when it has the `[[ noreturn ]]` attribute is not always possible.
+Since C++ has no type for it, `noreturn` in `extern(C++)` can be mangled as `void`, which is a common return type for `[[ noreturn ]]` functions in C++.
+When the return type isn't void, it can still be worked around by either ignoring the `[[ noreturn ]]` or adding wrapper code:
+
+```D
+extern(C++) int cppExit(); // returns int for some reason
+
+void main() {
+    writeln("hello world");
+    cppExit(); // won't return, but who cares?
+}
+
+// dExit can be used and it will interact with the type system correctly
+noreturn dExit() {
+    cast(noreturn) cppExit();
+}
+
+// the burden can also be put on the caller
+auto f = () => cast(noreturn) cppExit();
+```
+
+### Alternatives
+It has been proposed that an attribute such as `@noreturn` could be a simpler solution for specifying that a function does not return.
+Many C compilers have a specific attribute like this, and C++11 even introduced a standard `[[noreturn]]` annotation.
+Since the DMD, LDC and GDC compilers each use a C-backend that supports the notion of "no return" functions, it could simply be unified in library code:
+```D
+version (LDC)
+    enum noreturn = ldc.attributes.llvmAttr("noreturn"));
+else version (GDC)
+    enum noreturn = gcc.attribute.attribute("noreturn");
+else version (DigitalMars)
+    enum noreturn = pragma(noreturn); // example syntax
+
+// usage:
+@noreturn void exit();
+```
+
+It has been claimed that since functions that do not return are rare, the added compiler complexity from adding a bottom type is not worth it.
+While in the short term it might seem like a good solution, such ad-hoc solutions don't scale well and eventually only add more complexity than ever.
```

Comment:
This sentence is formulated a bit more colloquially than I would expect in a DIP.